build_tools/check_backticks.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3 -u
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+
+"""Test script to check for invalid use of single-backticks."""
+
+__author__ = ["geetu040"]
+
+import argparse
+import ast
+import glob
+import re
+
+
+def find_py_files(folder_path):
+    """Find all Python files in a given folder path."""
+    return glob.glob(f"{folder_path}/**/*.py", recursive=True)
+
+
+def extract_docstrings(filename):
+    """Extract docstrings from a Python file."""
+    # create abstract syntax tree from the file
+    with open(filename) as f:
+        tree = ast.parse(f.read())
+
+    # walk through all nodes in the tree
+    docstrings = {}
+    for node in ast.walk(tree):
+        if (
+            isinstance(node, ast.Expr)
+            and isinstance(node.value, ast.Constant)
+            and isinstance(node.value.value, str)
+        ):
+            # if the node is an expression and
+            # its value is a constant and
+            # constant's value is a string
+            # the node represents a docstring
+            # See https://docs.python.org/3/library/ast.html#abstract-grammar
+            docstring = node.value.value
+            lineno = node.value.lineno
+            docstrings[lineno] = docstring
+
+    return docstrings
+
+
+def find_invalid_backtick_text(docstring):
+    """Find invalid backtick text in a docstring."""
+    # remove all multiple backticks to avoid interference
+    # we are looking only for invalid single-backtick
+    docstring = re.sub(r"`{2,}.*?`{2,}", "", docstring, flags=re.DOTALL)
```

Comment:
I don't think I agree with this pattern. It matches ```` ```abc`` ```` as well for example, where backticks are not matched both side. Is that acceptable?

Reply:
Here's an example of what I meant: https://regex101.com/r/CqxGSL/1

I'm not sure if it's foolproof or not, but hopefully you get the idea.

If this is complex and causing problems, I am perfectly happy with your previous version of separate expressions.

Reply:
> I don't think I agree with this pattern. It matches `` ```abc`` `` as well for example, where backticks are not matched both side. Is that acceptable?

I think that should be acceptable, because we basically want to remove all multiple-backtick to avoid inference when searching invalid single-backtick

Reply:
Isn't this script sufficient for *detection* though? I thought that was the primary purpose for use as pre-commit tool.

That is, *if* the script modifies the file, *then* there's something wrong with the backticks.

It would not fully fix the unmatched backticks, but it would modify the file, therefore purpose satisfied?

Reply:
If the goal is to detect possible (but not necessarily) invalid use of single backticks, I am mostly okay with current script. @geetu040 needs to add some changes to change the exit code, but otherwise it seems fine.

But then it's not suitable for pre-commit, as it'll fail every time unless you fix the issues. You can make it a manual stage hook and not run by default, but that'd limit the usefulness of the script.

Reply:
I'd argue it has its use then still, like the "create changelog" script. The latter is also buggy, but (until at least AI takes over here), it reduces the time to a complete changelog substantially.